source/slang/slang-ir-specialize-arrays.cpp

```diff
@@ -38,15 +41,55 @@ struct ArrayParameterSpecializationCondition : FunctionCallSpecializeCondition
     bool doesParamWantSpecialization(IRParam* param, IRInst* arg)
     {
         SLANG_UNUSED(arg);
-        return isStructTypeWithArray(param->getDataType());
+        if (isKhronosTarget(codeGenContext->getTargetReq()))
+            return isStructTypeWithArray(param->getDataType());
+        return false;
     }
+
+    bool doesParamTypeWantSpecialization(IRParam* param, IRInst* arg)
+    {
+        auto paramType = param->getDataType();
+        auto argType = arg->getDataType();
+        if (auto outTypeBase = as<IROutTypeBase>(paramType))
+        {
+            paramType = outTypeBase->getValueType();
+            SLANG_ASSERT(as<IRPtrTypeBase>(argType));
```

Comment:
I am little worried that this assertion may fail if the user shader is written incorrectly and the argument is not a pointer type.
Shouldn't it be an error or just return false like following?
```
if (!as<IRPtrTypeBase>(argType)) return false;
```

Reply:
If that happens, the IR is ill formed and that would be a bug in the front-end not properly capturing the error.

The purpose of the semantic checking is to make sure we never generate this kind of ill-formed code, so an ASSERT here is good enough.